### ISSUE TITLE
Messages not re-queuing after expired Heartbeat

### ISSUE THREAD

**Issue by AllTaken** [blehnen/DotNetWorkQueue#9]
The transport is MSSQL.
I have set up a consumer as follows:

```
HeartBeat.UpdateTime = "sec(*%10)";
HeartBeat.MonitorTime = TimeSpan.FromSeconds(15);
HeartBeat.Time = TimeSpan.FromSeconds(35);
MessageExpiration.Enabled = false;
```

I have an item in queue whose Heartbeat is expired:

```
SELECT 
	  A.[QueueID]
	  ,B.[Status] AS StatusStatus
	  ,C.HeartBeat
	  ,C.QueuedDateTime
	  ,C.[Status] AS MetaDataStatus
	  ,GETDATE() AS CurrentTime
  FROM [MessageQueue] A
LEFT JOIN
	[MessageQueueStatus] B
ON A.[QueueID] = B.[QueueID]
LEFT JOIN
	[MessageQueueMetaData] C
ON A.[QueueID] = C.[QueueID]
```

```
QueueID	StatusStatus	HeartBeat			QueuedDateTime		MetaDataStatus	CurrentTime
1	1		2019-12-19 16:40:31.153		2019-12-19 16:40:30.857	1		2019-12-20 13:06:14.140
```
All other MessageQueue tables are empty.

But when I start the consumer nothing happens with this item. If I enqueue a new item, the new item gets handled...

**Comment by blehnen** [comment]
It almost sounds like the monitor thread is encountering an error.  Do you have any logging setup?

Since the queue uses LibLog internally, you can get it to log to Serilog by doing this as an example:

using Serilog;

var log = new LoggerConfiguration()
.WriteTo.Console() //or file system, etc
.MinimumLevel.Debug()
.CreateLogger();
Log.Logger = log;

The above just needs to be before any queue container creation.

**Comment by AllTaken** [comment]
2019-12-20 16:24:13.682 +01:00 [DBG] [SQL] server difference is 1331.5173 MS
2019-12-20 16:24:13.679 +01:00 [DBG] Worker 1 created
2019-12-20 16:24:13.705 +01:00 [INF] Initializing with 1 workers
2019-12-20 16:24:13.736 +01:00 [INF] Queue started

**Comment by AllTaken** [comment]
And this is what it looks like if I enqueue another item:

2019-12-20 16:30:00.331 +01:00 [DBG] Worker 1 created
2019-12-20 16:30:00.355 +01:00 [DBG] [SQL] server difference is 1341.3627 MS
2019-12-20 16:30:00.363 +01:00 [INF] Initializing with 1 workers
2019-12-20 16:30:00.370 +01:00 [INF] Queue started
2019-12-20 16:30:10.108 +01:00 [INF] Scheduler time is 20-12-2019 15:30:11
2019-12-20 16:30:10.109 +01:00 [INF] Local time is 20-12-2019 15:30:10
2019-12-20 16:30:10.370 +01:00 [DBG] Worker 1 created
2019-12-20 16:30:10.372 +01:00 [INF] Initializing with 1 workers
2019-12-20 16:30:10.372 +01:00 [INF] Queue started
2019-12-20 16:30:13.037 +01:00 [DBG] My debug message from start of handler
2019-12-20 16:30:13.316 +01:00 [DBG] My debug message from end of handler

**Comment by blehnen** [comment]
If you run the SQL server sample producer/consumer, do you have the same issue?

I don't have access to a SQL server instance at the moment; I was able to test with SQLite and it appears to be working.  When it resets any heartbeats, it logs the entry like this:

Reset the status of N records that where outside of the heartbeat window of X seconds

**Comment by AllTaken** [comment]
It does strange things.

I enqueued 10 messages with the sample producer (option A).

Then I set a breakpoint in the handler in the sample consumer, ran it, and killed the process when it hit the breakpoint.
```

QueueID	StatusStatus	HeartBeat			QueuedDateTime		MetaDataStatus	CurrentTime
1	1		2019-12-20 20:46:25.260		2019-12-20 20:46:07.990	1		2019-12-20 21:48:03.613
2	1		2019-12-20 20:46:25.270		2019-12-20 20:46:08.007	1		2019-12-20 21:48:03.613
3	1		2019-12-20 20:46:25.270		2019-12-20 20:46:08.010	1		2019-12-20 21:48:03.613
4	1		2019-12-20 20:46:25.283		2019-12-20 20:46:08.033	1		2019-12-20 21:48:03.613
5	0		NULL				2019-12-20 20:46:08.037	0		2019-12-20 21:48:03.613
6	0		NULL				2019-12-20 20:46:08.040	0		2019-12-20 21:48:03.613
7	0		NULL				2019-12-20 20:46:08.043	0		2019-12-20 21:48:03.613
8	0		NULL				2019-12-20 20:46:08.050	0		2019-12-20 21:48:03.613
9	0		NULL				2019-12-20 20:46:08.050	0		2019-12-20 21:48:03.613
10	0		NULL				2019-12-20 20:46:08.070	0		2019-12-20 21:48:03.613
```

Since there are 4 workers, this left 4 messages with status "1".

I then waited for a couple of minutes and I then ran the consumer again and it did this:

```
[21:53:02 INF] Startup
[21:53:02 INF] Tracing:False Chaos:False Metrics:False Compression:True Encryption:True
[21:53:04 DBG] Worker 1 created
[21:53:04 DBG] [SQL] server difference is 1853.2699 MS
[21:53:04 INF] Initializing with 4 workers
[21:53:05 DBG] Worker 4 created
[21:53:05 DBG] Worker 3 created
[21:53:05 INF] Reset the status of 3 records that where outside of the heartbeat window of 35 seconds
[21:53:05 DBG] Worker 2 created
[21:53:05 INF] Queue started
Processing messages - press any key to stop
[21:53:05 INF] Scheduler time is 20-12-2019 20:53:07
[21:53:05 INF] Local time is 20-12-2019 20:53:05
[21:53:05 DBG] Worker 1 created
[21:53:05 INF] Initializing with 1 workers
[21:53:05 INF] Queue started
[21:53:05 INF] Processing message 3 - Processing time is 500
[21:53:05 INF] Processing message 2 - Processing time is 500
[21:53:05 INF] Processing message 5 - Processing time is 500
[21:53:05 INF] Processing message 1 - Processing time is 500
[21:53:06 INF] Message 3 complete
[21:53:06 INF] Message 1 complete
[21:53:06 INF] Message 2 complete
[21:53:06 INF] Message 5 complete
[21:53:06 INF] Processing message 7 - Processing time is 500
[21:53:06 INF] Processing message 6 - Processing time is 500
[21:53:06 INF] Processing message 9 - Processing time is 500
[21:53:06 INF] Processing message 8 - Processing time is 500
[21:53:07 INF] Message 7 complete
[21:53:07 INF] Message 6 complete
[21:53:07 INF] Message 8 complete
[21:53:07 INF] Message 9 complete
[21:53:07 INF] Processing message 10 - Processing time is 500
[21:53:07 INF] Message 10 complete
```

So it did retry 3 of the stale messages, but 1 is still there


If I restart the consumer nothing happens.

I repeated this experiment a couple of times, and it seems "random" which messages it will reset. I have seen anywhere from 0 to 3.

If I take the heartbeat time from one of the messages that works, and copy it to the one that doesn't it will then work as well.

So I'm guessing something is wrong in the datetime arithmetic somewhere.

**Comment by blehnen** [comment]
Ok, this sounds like a bug with the SQL server transport.  We use the PostGresSQL and SQLite transports in production, but we don't use the SQL server or Redis transports at all, which is probably why this has never been reported internally.

I will take a look at this; however, I am getting ready to leave for the holidays and won't have a chance to investigate until January.

Since all transports use shared or similar logic, it might be an issue with this SQL statement -

 $"select {TableNameHelper.MetaDataName}.queueid, heartbeat, headers from {TableNameHelper.MetaDataName} with (updlock, readpast, rowlock) inner join {TableNameHelper.QueueName} on {TableNameHelper.QueueName}.queueid = {TableNameHelper.MetaDataName}.queueid where status = @status and heartbeat is not null and (DATEDIFF(SECOND, heartbeat, GETUTCDATE()) > @time)");

**Comment by AllTaken** [comment]
I was wrong.
The issue has to do with the ResetHeartBeat query generated for SQL Server.
The following query is run:
```

"update sampleQueueMetaData with (updlock, readpast, rowlock) set status = @Status, heartbeat = null where queueID = @QueueID and status = @SourceStatus and HeartBeat = @HeartBeat"

@QueueID = 11
@SourceStatus = 1
@Status = 0
@HeartBeat = {20-12-2019 20:55:50}
```
I believe the issue has to do with the following line in `DotNetWorkQueue.Transport.RelationalDatabase.Basic.CommandPrepareHandler`:

`65            param.DbType = DbType.DateTime2;`

The precision of `DateTime2` is not the same as `DateTime`, and while the parameter is set to `DateTime2` the generated column in the SQL Server database is `DateTime`.

If I change the parameter to `DateTime` it seems to work. I can't say if that is a viable solution though; I have no idea how the other databases might react. Maybe changing the SQL Server column to `DateTime2` would work?

**Comment by AllTaken** [comment]
Having done a few quick tests, changing the column datatype to datetime2 also seems to fix this issue.

Changing the ResetHeartBeat query for SQL Server to:
`"update sampleQueueMetaData with (updlock, readpast, rowlock) set status = @Status, heartbeat = null where queueID = @QueueID and status = @SourceStatus and HeartBeat = CAST(@HeartBeat AS DATETIME)"`

Also appears to fix the issue.

I don't know which if any of these would be better.

For now I wish you happy holidays!

**Comment by blehnen** [comment]
Thanks for your assistance;  The integration tests do pick up this issue.  Unfortunately, they where disabled on my team city server.  I'm not sure how/why that happened.

I've checked in the fix above - switching to DateTime seems to fix the issue, and I've noticed no side affects with the other transports so far.  The full integration tests take around 8 hours to run on my rather outdated hardware - Assuming everything goes ok, I'm hoping to publish a new version on Monday, before I leave town.
